dask/dataframe/multi.py

```diff
@@ -412,42 +412,50 @@ def single_partition_join(left, right, **kwargs):
     kwargs["empty_index_dtype"] = meta.index.dtype
     kwargs["categorical_columns"] = meta.select_dtypes(include="category").columns
 
-    name = "merge-" + tokenize(left, right, **kwargs)
-
     if right.npartitions == 1 and kwargs["how"] in allowed_left:
-        right_key = first(right.__dask_keys__())
-        dsk = {
-            (name, i): (apply, merge_chunk, [left_key, right_key], kwargs)
-            for i, left_key in enumerate(left.__dask_keys__())
-        }
+        joined = left.map_partitions(
+            merge_chunk,
+            right,
+            meta=meta,
+            enforce_metadata=False,
+            transform_divisions=False,
+            align_dataframes=False,
+            token="merge",  # NOTE: misleadingly, this is actually the name
+            **kwargs,
+        )
+
         if use_left:
-            divisions = left.divisions
+            joined.divisions = left.divisions
         elif use_right and len(right.divisions) == len(left.divisions):
-            divisions = right.divisions
+            joined.divisions = right.divisions
         else:
-            divisions = [None for _ in left.divisions]
+            joined.divisions = [None for _ in left.divisions]
 
     elif left.npartitions == 1 and kwargs["how"] in allowed_right:
-        left_key = first(left.__dask_keys__())
-        dsk = {
-            (name, i): (apply, merge_chunk, [left_key, right_key], kwargs)
-            for i, right_key in enumerate(right.__dask_keys__())
-        }
+        joined = right.map_partitions(
+            lambda right, left, **kwargs: merge_chunk(left, right, **kwargs),
+            left,
```

Comment:
I feel like I've heard that we try to avoid inlining lambda functions in the code base, but I can't remember why. It seems like you could just use `dd.map_partitions` and achieve the same outcome thought right?

```suggestion
        joined = map_partitions(
            merge_chunk,
            left,
            right,
```